README.md
# hasky-interpreter
se tiver stack, que facilitará para baixar o pacote Parsec:
```
stack setup && \
stack build
```
para rodar com stack:
```
stack exec hasky-interpreter-exe
```
se não tiver, então instala como quiser o Text.Parsec >= 3.1.13.0 e roda o `app/Main.hs`

# Como usar
entra com o programa.
exemplos:
```
Bem-vindo ao hasky-interpreter

declare x=150 in declare y = 200 in { while x < 300 do { x:= x+1; y := y - 1};print y}
```
```
Bem-vindo ao hasky-interpreter

declare x=150 in declare y = 200 in { if y = x then { x:= x+1; y := y - 1} else x:=2 ;print y}
```

```
Bem-vindo ao hasky-interpreter

declare x=150 in declare y = 150 in { if y = x then { x:= x+1; y := y - 1} else x:=2 ;print y+x}
```

```
Bem-vindo ao hasky-interpreter

declare x=150 in declare y = 200 in { if y = x then { x:= x+1; y := y - 1} else x:=2 ;print y+x}
```

ChangeLog.md
# Changelog for hasky-interpreter

## Unreleased changes


back-end.hs
data Exp = Constant Int
    | Variable String
    | Plus Exp Exp
    | Minus Exp Exp
    | Greater Exp Exp
    | Times Exp Exp
    deriving Show

data Com =  Assign String Exp
    |  Seq Com Com
    |  Cond Exp Com Com
    |  While Exp Com
    |  Declare String Exp Com
    |  Print Exp
    deriving Show

type SymbolTable = [(String, Int)]

tableset :: String -> Int -> SymbolTable -> SymbolTable
tableset sym new [] = [(sym, new)]
tableset sym new ((tablesym, old):vs) =
    if tablesym == sym
        then ((sym, new):vs)
        else ((tablesym, old):(tableset sym new vs))

tableget :: String -> SymbolTable -> Int
tableget _ [] = 0
tableget sym ((tablesym, val):vs) =
    if tablesym == sym
        then val
        else tableget sym vs

--define this as a monad
newtype M a = StOut (SymbolTable -> (a ,  SymbolTable , String))

unStOut :: M a -> SymbolTable -> (a, SymbolTable, String)
unStOut (StOut f) = f

applyToFirst :: (a->a1) -> (a,b,c) -> (a1,b,c)
applyToFirst f (a,b,c) = (f a, b, c)

instance Functor M where
    fmap func (StOut innerfunc) = StOut (applyToFirst func . innerfunc)

compose func1 func2 table = (resfunc resval, restable, resstring)
    where (resval, restable, resstring) = func2 table
          (resfunc,_,_) = func1 table

instance Applicative M where
    pure a = StOut (\x -> (a, x, []))
    (<*>) (StOut func1) (StOut func2) = StOut(compose func1 func2)

first :: (a,b,c) -> a
first (a,_,_) = a

instance Monad M where
    return a = pure a
    (>>=) (StOut innerfunc) func = StOut (\x ->
        let (r1, s1, o1) = innerfunc x
            StOut(f) = func r1
            (r2, s2, o2) = f s1
        in (r2, s2, o1 ++ o2))


unwrap :: M a -> a
unwrap (StOut f) = first $ f []

eval :: Exp -> SymbolTable -> M Int
eval expr table = case expr of
    Constant n -> return n
    Variable x -> return (tableget x table)
    Plus e1 e2 -> return $ unwrap (eval e1 table) + unwrap (eval e2 table)
    Minus e1 e2 -> return $ unwrap (eval e1 table) - unwrap (eval e2 table)
    Greater e1 e2 -> if unwrap (eval e1 table) > unwrap (eval e2 table)
                        then return 1
                        else return 0
    Times e1 e2 -> return $ unwrap (eval e1 table) * unwrap (eval e2 table)

exec :: Com -> SymbolTable -> M SymbolTable
exec stm table = case stm of
    Assign name expr -> return $ tableset name (unwrap $ eval expr table) table
    Seq com1 com2 -> do
        table1 <- exec com1 table
        table2 <- exec com2 table1
        return table2
    Cond expr com1 com2 ->
        if 0 /= (unwrap $ eval expr table)
        then exec com1 table
        else exec com2 table
    While expr com -> do
        if 0 /= (unwrap $ eval expr table)
        then exec stm (unwrap $ exec com table)
        else return table
    Declare name expr com -> do
        --special insertion at the front guarantees local variable properties
        table1 <- exec com ((name, unwrap $ eval expr table):table)
        return (removefirst table1)
            where removefirst ((nam, val):vs) = if nam == name
                                              then vs
                                              else (nam,val):(removefirst vs)
    Print expr -> StOut (\st -> ([], st, (show $ unwrap $ eval expr table) ++ "\n"))

test :: String
test = let
    StOut f = exec (Seq (Assign "x" (Plus (Constant 5)  (Constant 2))) (Print (Variable "x"))) []
    (_,_,str) = (f [])
    in str

src/Back.hs
module Back
  (
    Exp (Constant, Variable, Plus, Minus, Greater, Less, Equal, Times, Divide)
  , Com (Assign, Seq, Cond, While, Declare, Print)
  , interp
  ) where

data Exp = Constant Int
         | Variable String
         | Plus Exp Exp
         | Minus Exp Exp
         | Greater Exp Exp
         | Less Exp Exp
         | Equal Exp Exp
         | Times Exp Exp
         | Divide Exp Exp
         deriving Show

data Com = Assign Exp Exp
         | Seq Com Com
         | Cond Exp Com Com
         | While Exp Com
         | Declare Exp Exp Com
         | Print Exp
         deriving Show

type SymbolTable = [(String, Int)]

tableset :: Exp -> Int -> SymbolTable -> SymbolTable
tableset s@(Variable sym) new [] = [(sym, new)]
tableset s@(Variable sym) new ((tablesym, old):vs) =
  if tablesym == sym
  then ((sym, new):vs)
  else ((tablesym, old):(tableset s new vs))

tableget :: Exp -> SymbolTable -> Int
tableget _ [] = 0
tableget s@(Variable sym) ((tablesym, val):vs) =
  if tablesym == sym
  then val
  else tableget s vs

--define this as a monad
newtype M a = StOut (SymbolTable -> (a ,  SymbolTable , String))

unStOut :: M a -> SymbolTable -> (a, SymbolTable, String)
unStOut (StOut f) = f

applyToFirst :: (a->a1) -> (a,b,c) -> (a1,b,c)
applyToFirst f (a,b,c) = (f a, b, c)

instance Functor M where
    fmap func (StOut innerfunc) = StOut (applyToFirst func . innerfunc)

compose func1 func2 table = (resfunc resval, restable, resstring)
    where (resval, restable, resstring) = func2 table
          (resfunc,_,_) = func1 table

instance Applicative M where
    pure a = StOut (\x -> (a, x, []))
    (<*>) (StOut func1) (StOut func2) = StOut(compose func1 func2)

first :: (a,b,c) -> a
first (a,_,_) = a

instance Monad M where
    return a = pure a
    (>>=) (StOut innerfunc) func = StOut (\x ->
        let (r1, s1, o1) = innerfunc x
            StOut(f) = func r1
            (r2, s2, o2) = f s1
        in (r2, s2, o1 ++ o2))


unwrap :: M a -> a
unwrap (StOut f) = first $ f []

eval :: Exp -> SymbolTable -> M Int
eval expr table = case expr of
    Constant n -> return n
    v@(Variable x) -> return (tableget v table)
    Plus e1 e2 -> return $ unwrap (eval e1 table) + unwrap (eval e2 table)
    Minus e1 e2 -> return $ unwrap (eval e1 table) - unwrap (eval e2 table)
    Greater e1 e2 -> if unwrap (eval e1 table) > unwrap (eval e2 table)
                     then return 1
                     else return 0
    Less e1 e2 -> if unwrap (eval e1 table) < unwrap (eval e2 table)
                  then return 1
                  else return 0
    Equal e1 e2 -> if unwrap (eval e1 table) == unwrap (eval e2 table)
                   then return 1
                   else return 0
    Times e1 e2 -> return $ unwrap (eval e1 table) * unwrap (eval e2 table)
    Divide e1 e2 -> return $ unwrap (eval e1 table) `div` unwrap (eval e2 table)

exec :: Com -> SymbolTable -> M SymbolTable
exec stm table = case stm of
  Assign name expr -> return $ tableset name (unwrap $ eval expr table) table
  Seq com1 com2 -> do
    table1 <- exec com1 table
    table2 <- exec com2 table1
    return table2
  Cond expr com1 com2 ->
    if 0 /= (unwrap $ eval expr table)
    then exec com1 table
    else exec com2 table
  While expr com -> do
    if 0 /= (unwrap $ eval expr table)
      then exec stm (unwrap $ exec com table)
      else return table
  Declare (Variable name) expr com -> do
    --special insertion at the front guarantees local variable properties
    table1 <- exec com ((name, unwrap $ eval expr table):table)
    return (removefirst table1)
      where removefirst ((nam, val):vs) = if nam == name
                                          then vs
                                          else (nam,val):(removefirst vs)
            removefirst [] = []
  Print expr -> StOut (\st -> ([], st, (show $ unwrap $ eval expr table)))


interp a = unStOut (exec a []) []

test :: String
test = let
    StOut f = exec (Seq (Assign (Variable "x") (Plus (Constant 5)  (Constant 2))) (Print (Variable "x"))) []
    (_,_,str) = (f [])
    in str


src/front-end.hs
import Text.ParserCombinators.Parsec
import Text.Parsec (ParsecT)
import Text.Parsec.Prim (Stream)
import PolyConc
import Data.Functor.Identity (Identity)

-- EXPRESSOES
-- <relop> ::= > | < | =
relop = (char '>')
        <|> (char '<')
        <|> char '='
        <?> "relop"

-- <mulop> ::= * | /
mulop = (char '*')
        <|> char '/'
        <?> "mulop"

-- <addop> ::= + | -
addop = (char '+')
        <|> char '-'
        <?> "addop"

-- <digit> ::= 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
-- digit ja esta definido no parsec

-- <digiti> ::= <digit> | <digiti> <digiti>
digiti = many1 digit

-- <var> ::= <Identifier> que sera letter ++ alphanum*
var = (:) <$> letter <*> many alphaNum

-- <factor> ::= <var> | <digiti> | (<expr>)
factor :: ParsecT [Char] u Identity [Char]
factor =
  var
  <|> digiti
  <|> (polyConc) <$> (string "(" <* spaces) <*> expr <*> (spaces *> string ")")
  <?> "factor"

-- <term> ::= <term> <mulop> <factor> | <factor>
-- eliminado recursao a' esquerda:
-- <term> ::= <factor> <term'>
-- <term'> ::= <mulop> <factor> <term'> | vazio
term = (++) <$> (factor) <*> term'
  where
    term' = try (
      (polyConc) <$> (spaces *> mulop <* spaces) <*> factor <*> term'
      )
      <|> string ""
      <?> "term"

-- <expr> ::= <expr> <addop> <term> | <term>

-- <expr> ::=  <term> <expr'>
-- <expr'> ::= <addop> <term> <expr'> | vazio
expr = (++) <$> (term) <*> expr'
  where
    expr' = try (
      (polyConc) <$> (spaces *> addop <* spaces) <*> term <*> expr'
      )
      <|> string ""
      <?> "expr"
  
  
-- <rexp> ::= <rexp> <relop> <expr> | <expr>

-- <rexp> ::= <expr> <rexp'>
-- <rexp'> ::= <relop> <expr> <rexp'> | vazio
rexp = (++) <$> (expr) <*> rexp'
  where
    rexp' = try (
      (polyConc) <$> (spaces *> relop <* spaces) <*> expr <*> rexp'
      )
      <|> string ""
      <?> "rexp"


-- COMANDOS
-- <printe> ::= print <rexp>
printe = (++) <$> string "print" <*> aux
  where aux = (polyConc) <$> space <* spaces <*> rexp

-- <declare> ::= declare <identif> = <rexp> in <com>
-- <identif> sera meu <var>
declare = (++) <$> string "declare" <*> aux
  where
    aux = (polyConc) <$> space <* spaces <*> var <*> (spaces *> string "=" <* spaces) <*> rexp <*> (space <* spaces) <*> string "in" <*> (space <* spaces) <*> com

-- <while> := while <rexp> do <com>
while = (++) <$> string "while" <*> aux
  where
    aux = (polyConc) <$> space <* spaces <*> rexp <*> (space <* spaces) <*> string "do" <*> (space <* spaces) <*> com

-- <cond> := if <rexp> then <com> else <com>
cond = (++) <$> string "if" <*> aux
  where
    aux = (polyConc) <$> space <* spaces <*> rexp <*> (space <* spaces) <*> string "then" <*> (space <* spaces) <*> com <*> (space <* spaces) <*> string "else" <*> (space <* spaces) <*> com

-- <seqv> ::= { <com> ; <com> }
seqv = (:) <$> char '{' <* spaces <*> aux
  where
    aux = (polyConc) <$> com <*> (spaces *> char ';' <* spaces) <*> com <*> (spaces *> char '}')

-- <assign> ::= <identif> := <rexp>
assign = (++) <$> var <* spaces <*> aux
  where aux = (polyConc) <$> (string ":=" <* spaces) <*> rexp

-- <com> ::= <assign> | <seqv> | <cond> | <while> | <declare> | <printe>
com = try assign <|> try seqv <|> try cond <|> try while <|> try declare <|> printe <?> "com"


test :: IO ()
test = do
  putStrLn "TESTANDO..."

  -- rexp
  let testN = 1 :: Integer
  putStrLn "\ntestando rexp"
  putStr "teste..." >> print testN
  let ret = parse rexp "" ""
  case (ret) of
    Right e -> putStrLn "ERRO"
    Left e -> putStrLn "OK"
  testN <- return (succ testN)
  
  putStr "teste..." >> print testN
  let ret = parse rexp "" "a > b + 2 * a / 1 /a"
  case (ret) of
    Right e -> if (e == "a>b+2*a/1/a")
                  then putStrLn "OK" else putStr "ERRO" >> print e >> putStrLn ""
    Left e -> putStrLn "ERRO"
  testN <- return (succ testN)
  
  -- printe
  let testN = 1 :: Integer
  putStrLn "\ntestando printe"
  putStr "teste..." >> print testN
  let ret = parse printe "" ""
  case (ret) of
    Right e -> putStrLn "ERRO" >> print e
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse printe "" "print     2"
  case (ret) of
    Right e -> if (e == "print 2")
               then putStrLn "OK" else putStr "ERRO" >> print e >> putStrLn ""
    Left e -> putStrLn "ERRO"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse printe "" "print2"
  case (ret) of
    Right e -> putStrLn "ERRO" >> print e
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse printe "" "print(2)"
  case (ret) of
    Right e -> putStrLn "ERRO" >> print e
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse printe "" "print     (2 * 2)"
  case (ret) of
    Right e -> if (e == "print (2*2)")
               then putStrLn "OK" else putStr "ERRO - " >> print e >> putStrLn ""
    Left e -> putStrLn "ERRO"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse printe "" "print     (   2 * 2 + 3) 1"
  case (ret) of
    Right e -> if (e == "print (2*2+3)")
               then putStrLn "OK" else putStr "ERRO - " >> print e >> putStrLn ""
    Left e -> putStrLn "ERRO"
  testN <- return (succ testN)

  -- testando declare
  let testN = 1 :: Integer
  putStrLn "\ntestando declare"
  putStr "teste..." >> print testN
  let ret = parse declare "" ""
  case (ret) of
    Right e -> putStrLn "ERRO"
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse declare "" "declare a = 2 in print 2"
  case (ret) of
    Right e -> if (e == "declare a=2 in print 2")
               then putStrLn "OK" else putStr "ERRO - " >> print e >> putStrLn ""
    Left e -> putStrLn "ERRO"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse declare "" "declare    a=2    in print    2      "
  case (ret) of
    Right e -> if (e == "declare a=2 in print 2")
               then putStrLn "OK" else putStr "ERRO - " >> print e >> putStrLn ""
    Left e -> putStrLn "ERRO"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse declare "" "declare    a=2inprint    2      "
  case (ret) of
    Right e -> putStrLn "ERRO" >> print e
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse declare "" "declare    a=2 inprint    2      "
  case (ret) of
    Right e -> putStrLn "ERRO" >> print e
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse declare "" "declare    a=2in print    2      "
  case (ret) of
    Right e -> putStrLn "ERRO" >> print e >> putStrLn ""
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  -- testando assign
  let testN = 1 :: Integer
  putStrLn "\ntestando assign"
  putStr "teste..." >> print testN
  let ret = parse assign "" ""
  case (ret) of
    Right e -> putStrLn "ERRO" >> print e
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse assign "" "a:=2"
  case (ret) of
    Right e -> if (e == "a:=2")
               then putStrLn "OK" else putStrLn "ERRO - " >> print e >> putStrLn ""
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse assign "" "a :=2"
  case (ret) of
    Right e -> if (e == "a:=2")
               then putStrLn "OK" else putStrLn "ERRO - " >> print e >> putStrLn ""
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse assign "" "a:= 2"
  case (ret) of
    Right e -> if (e == "a:=2")
               then putStrLn "OK" else putStrLn "ERRO - " >> print e >> putStrLn ""
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse assign "" "a := 2"
  case (ret) of
    Right e -> if (e == "a:=2")
               then putStrLn "OK" else putStrLn "ERRO - " >> print e >> putStrLn ""
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  -- testando while
  let testN = 1 :: Integer
  putStrLn "\ntestando while"
  putStr "teste..." >> print testN
  let ret = parse assign "" ""
  case (ret) of
    Right e -> putStrLn "ERRO" >> print e
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse assign "" "while (2) do a:=2"
  case (ret) of
    Right e -> if (e == "while (2) do a:=2")
               then putStrLn "OK" else putStrLn "ERRO - " >> print e >> putStrLn ""
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  -- testando cond
  let testN = 1 :: Integer
  putStrLn "\ntestando cond"
  putStr "teste..." >> print testN
  let ret = parse cond "" ""
  case (ret) of
    Right e -> putStrLn "ERRO"
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse cond "" "if (2) then a:=2 else a:=3"
  case (ret) of
    Right e -> if (e == "if (2) then a:=2 else a:=3") 
               then putStrLn "OK" else putStrLn "ERRO - " >> print e >> putStrLn ""
    Left e -> putStrLn "ERRO"
  testN <- return (succ testN)
  
  putStr "teste..." >> print testN
  let ret = parse cond "" "if (2) then   a:=2   else   a:=3"
  case (ret) of
    Right e -> if (e == "if (2) then a:=2 else a:=3") 
               then putStrLn "OK" else putStrLn "ERRO - " >> print e >> putStrLn ""
    Left e -> putStrLn "ERRO"
  testN <- return (succ testN)

  putStr "teste..." >> print testN
  let ret = parse cond "" "if (2)then   a:=2   else   a:=3"
  case (ret) of
    Right e -> putStr "ERRO - " >> print e >> putStrLn ""
    Left e -> putStrLn "OK"
  testN <- return (succ testN)

  -- testando programa da especificacao

  putStrLn "\ntestando programa da especificacao"
  let ret = parse com "" "declare x=150 in declare y = 200 in { while x > 0 do { x:= x-1; y := y - 1};print y}"
  case (ret) of
    Right e -> if (e == "declare x=150 in declare y=200 in {while x>0 do {x:=x-1;y:=y-1};print y}")
      then putStrLn "OK" else putStr "ERRO - " >> print e >> putStrLn ""
    Left e -> putStrLn "ERRO"

  return ()


src/Front.hs
module Front
  (
    compute
  ) where

import Text.ParserCombinators.Parsec
import Back

{-
-- EXPRESSOES
-- <relop> ::= > | < | =
relop = (char '>')
        <|> (char '<')
        <|> char '='
        <?> "relop"

-- <mulop> ::= * | /
mulop = (char '*')
        <|> char '/'
        <?> "mulop"

-- <addop> ::= + | -
addop = (char '+')
        <|> char '-'
        <?> "addop"
-}
-- <digit> ::= 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
-- digit ja esta definido no parsec

-- <digiti> ::= <digit> | <digiti> <digiti>
digiti = (Constant . read) <$> (many1 digit)

-- <var> ::= <Identifier> que sera letter ++ alphanum*
var = (Variable . show) <$> ((:) <$> letter <*> many alphaNum)

-- <factor> ::= <var> | <digiti> | (<expr>)
factor =
  var
  <|> digiti
  <|> (string "(" *> spaces) *> expr <* (spaces *> string ")")
  <?> "factor"

-- <term> ::= <term> '*' <factor> | <tem> '/' <factor | <factor>
-- eliminado recursao a' esquerda:
-- <term> ::= <factor> <term'>
-- <term'> ::= '*' <factor> <term'> | '/' <factor> <term'> | vazio
term = try (Times <$> factor <*> termF)
       <|> try (Divide <$> factor <*> termD)
       <|> factor
       where
         termF = try (Times <$> (spaces *> char '*' *> spaces *> factor) <*> termF)
                 <|> try (Divide <$> (spaces *> char '*' *> spaces *> factor) <*> termD)
                 <|> (spaces *> char '*' *> spaces *> factor)
         termD = try (Times <$> (spaces *> char '/' *> spaces *> factor) <*> termF)
                 <|> try (Divide <$> (spaces *> char '/' *> spaces *> factor) <*> termD)
                 <|> (spaces *> char '/' *> spaces *> factor)

-- <expr> ::= <expr> <addop> <term> | <term>
-- <expr> ::=  <term> <expr'>
-- <expr'> ::= <addop> <term> <expr'> | vazio
expr = try (Plus <$> term <*> exprP)
       <|> try (Minus <$> term <*> exprM)
       <|> term
       where
         exprP = try (Plus <$> (spaces *> char '+' *> spaces *> term) <*> exprP)
                 <|> try (Minus <$> (spaces *> char '+' *> spaces *> term) <*> exprM)
                 <|> (spaces *> char '+' *> spaces *> term)
         exprM = try (Plus <$> (spaces *> char '-' *> spaces *> term) <*> exprP)
                 <|> try (Minus <$> (spaces *> char '-' *> spaces *> term) <*> exprM)
                 <|> (spaces *> char '-' *> spaces *> term)
  
-- <rexp> ::= <rexp> <relop> <expr> | <expr>

-- <rexp> ::= <expr> <rexp'>
-- <rexp'> ::= <relop> <expr> <rexp'> | vazio
rexp = try (Greater <$> expr <*> rexpG)
       <|> try (Less <$> expr <*> rexpL)
       <|> try (Equal <$> expr <*> rexpE)
       <|> expr
       where
         rexpG = try (Greater <$> (spaces *> char '>' *> spaces *> expr) <*> rexpG)
                 <|> try (Less <$> (spaces *> char '>' *> spaces *> expr) <*> rexpL)
                 <|> try (Equal <$> (spaces *> char '>' *> spaces *> expr) <*> rexpE)
                 <|> (spaces *> char '>' *> spaces *> expr)
         rexpL = try (Greater <$> (spaces *> char '<' *> spaces *> expr) <*> rexpG)
                 <|> try (Less <$> (spaces *> char '<' *> spaces *> expr) <*> rexpL)
                 <|> try (Equal <$> (spaces *> char '<' *> spaces *> expr) <*> rexpE)
                 <|> (spaces *> char '<' *> spaces *> expr)
         rexpE = try (Greater <$> (spaces *> char '=' *> spaces *> expr) <*> rexpG)
                 <|> try (Less <$> (spaces *> char '=' *> spaces *> expr) <*> rexpL)
                 <|> try (Equal <$> (spaces *> char '=' *> spaces *> expr) <*> rexpE)
                 <|> (spaces *> char '=' *> spaces *> expr)


-- COMANDOS
-- <printe> ::= print <rexp>
printe = (Print) <$> (string "print" *> space *> spaces *> rexp)

-- <declare> ::= declare <identif> = <rexp> in <com>
-- <identif> sera meu <var>
declare = (Declare) <$> ((string "declare") *> space *> spaces *> var) <*> (spaces *> string "=" *> spaces *> rexp) <*> (space *> spaces *> string "in" *> space *> spaces *> com)

-- <while> := while <rexp> do <com>
while = (While) <$> ((string "while") *> space *> spaces *> rexp) <*> (space *> spaces *> string "do" *> space *> spaces *> com)

-- <cond> := if <rexp> then <com> else <com>
cond = (Cond) <$> ((string "if") *> space *> spaces *> rexp) <*> (space <* spaces *> string "then" *> space *> spaces *> com) <*> (space *> spaces *> string "else" *> space *> spaces *> com)

-- <seqv> ::= { <com> ; <com> }
seqv = (Seq) <$> ((char '{') *> spaces *> com) <*> (spaces *> char ';' *> spaces *> com) <* (spaces *> char '}')

-- <assign> ::= <identif> := <rexp>
assign = (Assign) <$> var <*> (spaces *> string ":=" *> spaces *> rexp)

-- <com> ::= <assign> | <seqv> | <cond> | <while> | <declare> | <printe>
com = try assign <|> try seqv <|> try cond <|> try while <|> try declare <|> printe <?> "com"


compute = c' . parse com ""
  where c' (Right v) = print $ interp v
        c' (Left v) = print v

test :: IO ()
test = do
  putStrLn "DEVERIA DAR 50"
  compute "declare x=150 in declare y = 200 in { while x < 300 do { x:= x+1; y := y - 1};print y}"

  putStrLn "\n\nDEVERIA DAR 200"
  compute "declare x=150 in declare y = 200 in { if y = x then { x:= x+1; y := y - 1} else x:=2 ;print y}"

  putStrLn "\n\nDEVERIA DAR 149"
  compute "declare x=150 in declare y = 150 in { if y = x then { x:= x+1; y := y - 1} else x:=2 ;print y}"

  putStrLn "\n\nDEVERIA DAR 300"
  compute "declare x=150 in declare y = 150 in { if y = x then { x:= x+1; y := y - 1} else x:=2 ;print y+x}"

  putStrLn "\n\nDEVERIA DAR 202"
  compute "declare x=150 in declare y = 200 in { if y = x then { x:= x+1; y := y - 1} else x:=2 ;print y+x}"
